File: leetcode/14_trees/235_Lowest_Common_Ancestor_of_a_Binary_Search_Tree.go
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	var (
		l int
		r int
	)

	if p.Val > q.Val {
		l = q.Val
		r = p.Val
	} else {
		l = p.Val
		r = q.Val
	}

	if root.Val >= l && root.Val <= r {
		return root
	} else if root.Val < l {
		return lowestCommonAncestor(root.Right, p, q)
	} else {
		return lowestCommonAncestor(root.Left, p, q)
	}
}

File: leetcode/14_trees/235_Lowest_Common_Ancestor_of_a_Binary_Search_Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if not root:
            return None

        if p.val > q.val:
            l = q.val
            r = p.val
        else:
            l = p.val
            r = q.val

        if root.val >= l and root.val <= r:
            return root
        elif root.val < l:
            return self.lowestCommonAncestor(root.right, p, q)
        else:
            return self.lowestCommonAncestor(root.left, p, q)
